### PR TITLE
feature: stats of cri manager

### PR DESCRIPTION
**Pull request by Starnop**
Signed-off-by: Starnop <starnop@163.com>

### Ⅰ. Describe what this PR did
This PR implement ContainerStats(), ListContainerStats() and ImageFsInfo() of CRI manager.

All of these interfaces are used to retrieve the metrics of containers or image file system info.

### Ⅱ. Does this pull request fix one issue?

None

### Ⅳ. Describe how to verify it
Use crictl tools:
```
# crictl stats
DEBU[0000] ListContainerStatsRequest: &ListContainerStatsRequest{Filter:&ContainerStatsFilter{Id:,PodSandboxId:,LabelSelector:map[string]string{},},} 

......

CONTAINER           CPU %               MEM                 DISK                INODES
0bbc13aa25d81       0.59                30.14MB             45.06kB             14
15eef79448135       0.00                286.7kB             24.58kB             9
2e257f09a2260       0.09                9.581MB             24.58kB             6
476ae4d251fe6       0.00                290.8kB             24.58kB             9
574b07cfd946c       0.03                10.21MB             8.192kB             3
587856b374711       0.44                11.49MB             24.58kB             10
6df66169f907a       1.18                37.26MB             61.44kB             19
725a1d78af2b0       0.00                290.8kB             24.58kB             9
8b44f5423bd64       0.02                6.955MB             12.29kB             4
a1f9ccae689ed       0.00                274.4kB             24.58kB             9
a42c9875b9e82       1.13                233.7MB             36.86kB             12
c4321c743307d       0.00                225.3kB             24.58kB             9
cf2dda929ea0b       0.00                286.7kB             24.58kB             9
eb400edd89bb5       0.00                5.947MB             28.67kB             9

```

or use curl tools:

```
# curl -X GET http://127.0.0.1:10255/stats/summary
```

### CODE REVIEW

File: cri/config/config.go [hunk 1]
```diff
@@ -1,5 +1,12 @@
 package config
 
+const (
+	// K8sNamespace is the namespace we use to connect containerd when CRI is enabled.
+	K8sNamespace = "k8s.io"
+	// ContainerdAddr is the default unix socket address of containerd when CRI is enabled.
+	ContainerdAddr = "/var/run/containerd/containerd.sock"
```
Comment (allencloud) [inline cri/config/config.go]: Will this part in CRI directly communicate with containerd? 
If that we need to do a detailed plan to set up a long roadmap of this.
If this is a temporary work?
If we need to implement this is container manager first?
And something others.

Comment (Starnop) [reply cri/config/config.go]: @allencloud Actually, we need to do a detailed plan. : )

Comment (allencloud) [reply cri/config/config.go]: Talked with you offline, and I think we need update of this Part.

**Comment by YaoZengzeng** [comment]
LGTM, @allencloud @fuweid PTAL

I hope this PR could be merged ASAP, otherwise the k8s wouldn't get the monitor stats of container 😄

File: cri/v1alpha1/cri.go [hunk 2]
```diff
@@ -60,6 +60,15 @@ const (
 
 	// resolvConfPath is the abs path of resolv.conf on host or container.
 	resolvConfPath = "/etc/resolv.conf"
+
+	// statsCollectPeriod is the time duration we sync stats from containerd.
```
Comment (HusterWan) [inline cri/v1alpha1/cri.go]: %s/time/seconds/ WDYT?

File: vendor/vendor.json [hunk 3]
```diff
@@ -613,6 +613,12 @@
 			"revision": "e214231b295a8ea9479f11b70b35d5acf3556d9b",
 			"revisionTime": "2018-01-08T23:09:05Z"
 		},
+		{
+			"checksumSHA1": "1txgvJO9bgE703kK2EQ/NATy1Rg=",
+			"path": "github.com/coreos/go-systemd/dbus",
```
Comment (fuweid) [inline vendor/vendor.json]: There is only one dependency? I checked the file and found that there are a lot of unless and non-go files in the vendor. Could you mind to provide your govendor command when you add the dependency? Thanks

File: cri/v1alpha1/cri_utils.go [hunk 4]
```diff
@@ -770,3 +802,115 @@ func parseUserFromImageUser(id string) string {
 	// no group, just return the id
 	return id
 }
+
+func (c *CriManager) getContainerMetrics(meta *mgr.Container, stats *containerdtypes.Metric) (*runtime.ContainerStats, error) {
+	var usedBytes, inodesUsed uint64
+	cs := &runtime.ContainerStats{}
+
+	sn, err := c.SnapshotStore.Get(meta.ID)
```
Comment (HusterWan) [inline cri/v1alpha1/cri_utils.go]: snapshot id may not equals container ID later, comments for instension

File: vendor/vendor.json (hunk 3 above)
Comment (Starnop) [reply vendor/vendor.json]: of course, what I usually do when add new dependencies :
```
go get package
govendor add/update package
```

File: cri/v1alpha1/cri_utils.go (hunk 4 above)
Comment (Starnop) [reply cri/v1alpha1/cri_utils.go]: A comment has been added.

File: vendor/vendor.json (hunk 3 above)
Comment (fuweid) [reply vendor/vendor.json]: I afraid that it doesn't work. `go get` will fetch the latest code in your gopath. I think we should use `govendor fetch package@version`

File: daemon/mgr/snapshot.go [hunk 5]
```diff
@@ -0,0 +1,153 @@
...
+		usage, err := s.client.GetSnapshotUsage(context.Background(), info.Name)
+		if err != nil {
+			logrus.Errorf("failed to get usage for snapshot %q: %v", info.Name, err)
+			continue
+		}
+		sn.Size = uint64(usage.Size)
+		sn.Inodes = uint64(usage.Inodes)
+		s.store.Add(sn)
+	}
+	for _, sn := range s.store.List() {
```
Comment (fuweid) [inline daemon/mgr/snapshot.go]: Could you mind to add more information about this part? For example, about the delete logic. Why do you only delete the `out of date` snapshot?

File: cri/v1alpha1/cri.go [hunk 6]
```diff
@@ -697,12 +725,55 @@ func (c *CriManager) ContainerStatus(ctx context.Context, r *runtime.ContainerSt
 // ContainerStats returns stats of the container. If the container does not
 // exist, the call returns an error.
 func (c *CriManager) ContainerStats(ctx context.Context, r *runtime.ContainerStatsRequest) (*runtime.ContainerStatsResponse, error) {
-	return nil, fmt.Errorf("ContainerStats Not Implemented Yet")
+	containerID := r.GetContainerId()
+
+	container, err := c.ContainerMgr.Get(ctx, containerID)
```
Comment (fuweid) [inline cri/v1alpha1/cri.go]: Could we use a function to get the `cs`? If we have the function, we will reuse the function in the `ListContainerStats`. Don't repeat yourself.

File: cri/v1alpha1/cri.go [hunk 7]
```diff
@@ -952,5 +1023,25 @@ func (c *CriManager) RemoveImage(ctx context.Context, r *runtime.RemoveImageRequ
 
 // ImageFsInfo returns information of the filesystem that is used to store images.
 func (c *CriManager) ImageFsInfo(ctx context.Context, r *runtime.ImageFsInfoRequest) (*runtime.ImageFsInfoResponse, error) {
-	return nil, fmt.Errorf("ImageFsInfo Not Implemented Yet")
+	snapshots := c.SnapshotStore.List()
+	timestamp := time.Now().UnixNano()
+	var usedBytes, inodesUsed uint64
+	for _, sn := range snapshots {
+		// Use the oldest timestamp as the timestamp of imagefs info.
```
Comment (fuweid) [inline cri/v1alpha1/cri.go]: Could you add more information about this logic?

File: cri/v1alpha1/cri.go [hunk 8]
```diff
@@ -127,6 +143,18 @@ func NewCriManager(config *config.Config, ctrMgr mgr.ContainerMgr, imgMgr mgr.Im
 		return nil, fmt.Errorf("failed to create sandbox meta store: %v", err)
 	}
 
+	imageFSPath := imageFSPath(path.Join(config.HomeDir, "containerd/root"), defaultSnapshotterName)
```
[the remark this answers is not in the document]
Comment (Starnop) [reply cri/v1alpha1/cri.go]: A separate PR will be submitted to deal with similar things.

File: cri/v1alpha1/cri.go (hunk 6 above)
Comment (Starnop) [reply cri/v1alpha1/cri.go]: Maybe it is not necessary to do that, the function `getContainerMetrics` has been encapsulated, what has been repeated is just error handling.

File: daemon/mgr/snapshot.go [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+
+// Start starts the snapshots syncer.
+func (s *SnapshotsSyncer) Start() {
+	tick := time.NewTicker(s.syncPeriod)
+	go func() {
+		defer tick.Stop()
+		for {
+			err := s.Sync()
+			if err != nil {
+				logrus.Errorf("failed to sync snapshot stats: %v", err)
```
[the remark this answers is not in the document]
Comment (Starnop) [reply daemon/mgr/snapshot.go]: Done.

File: daemon/mgr/snapshot.go [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+
+// Get returns the snapshot with specified key. Returns error if the
+// snapshot doesn't exist.
+func (s *SnapshotStore) Get(key string) (Snapshot, error) {
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+	if sn, ok := s.snapshots[key]; ok {
+		return sn, nil
+	}
+	return Snapshot{}, fmt.Errorf("failed to get %q in snapshot store", key)
```
[the remark this answers is not in the document]
Comment (Starnop) [reply daemon/mgr/snapshot.go]: Done.

File: daemon/mgr/snapshot.go (hunk 5 above)
Comment (Starnop) [reply daemon/mgr/snapshot.go]: The purpose of this operation is to keep the synchronization with `client.SnapshotService`. A case:
when you remove a container,you also need to remove snapshot. However, `SnapshotStore` will not be notified. So wo need to delete from `SnapshotStore` that doesn't exist actually.

File: cri/v1alpha1/cri.go (hunk 6 above)
Comment (Starnop) [reply cri/v1alpha1/cri.go]: I have to admit it's more elegant. Thank you.

File: cri/v1alpha1/cri_utils.go [hunk 11]
```diff
@@ -17,10 +19,16 @@ import (
 	"github.com/alibaba/pouch/pkg/utils"
 	"github.com/go-openapi/strfmt"
```
Comment (fuweid) [inline cri/v1alpha1/cri_utils.go]: please move it into next group

File: daemon/mgr/snapshot.go [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+
+// SnapshotsSyncer syncs snapshot stats periodically.
+type SnapshotsSyncer struct {
+	store      *SnapshotStore
+	client     ctrd.APIClient
+	syncPeriod time.Duration
+}
+
+// NewSnapshotsSyncer creates a snapshot syncer.
+func NewSnapshotsSyncer(store *SnapshotStore, cli ctrd.APIClient, period time.Duration) *SnapshotsSyncer {
```
Comment (YaoZengzeng) [inline daemon/mgr/snapshot.go]: `NewSnapshotsSyncer ` will not be exported, rename it to `newSnapshotsSyncer ` would be better.

**Comment by YaoZengzeng** [comment]
Except the minor issue mentioned above, LGTM

**Comment by fuweid** [comment]
LGTM

**Pull request merged**